department-of-veterans-affairs/caseflow
# CASEFLOW-376 Put CAVC death-dismissal and reversals on-hold automatically

Resolves [CASEFLOW-376](https://vajira.max.gov/browse/CASEFLOW-376) and [CASEFLOW-411](https://vajira.max.gov/browse/CASEFLOW-411).

### Description
Enable CAVC types Death Dismissal and Straight Reversal.
* If mandate dates are not provided, create `MandateHoldTask` subtask.
* If mandate dates are provided, the created CAVC appeal should be ready for distribution.

#### User Stories 
As a CAVC Litigation Support user, if a mandate was not received for a CAVC Death Dismissal or Reversal I need the CAVC Death Dismissal case to be placed on hold, so that the case is not worked till the mandate is received. 

As a CAVC Litigation Support user, when I create a new AMA appeal for CAVC Death Dismissal & Reversal that has been issued mandate, I need that appeal to be sent directly to case distribution so that case can be distributed to a VLJ for a death dismissal or Reversal decision. 

### Acceptance Criteria
- [x] A new `MandateHoldTask` is created and placed on hold for 90 days
- [x] The new task is assigned to the CAVC Litigation Support team
- [x] The task is child of CAVC Task
- [x] The following actions are available on the task: 
   - [x] Put task on hold - presents on hold modal
- [x] Update documentation: [CAVC Remand](https://github.com/department-of-veterans-affairs/caseflow/wiki/CAVC-Remands)

Verify:
- [x] Verify a new appeal stream is created when the CAVC Litigation Support team selects Death Dismissal and clicks submit on the CAVC entry form
- [x] Verify a new appeal stream is created when the CAVC Litigation Support team selects Reversal and clicks submit on the CAVC entry form
- [x] Verify the new appeal stream is linked to the existing docket number at the Board
- [x] Verify the original and new appeal appears when searching for original docket #
- [x] Verify the remand information is captured in the new appeals remand section

Copied from [CASEFLOW-411](https://vajira.max.gov/browse/CASEFLOW-411):
- [x] CAVC Death Dismissals with mandate creates an AMA CAVC Death Dismissal appeal and marks the case as ready for distribution.
- [x] CAVC Reversal with mandate creates an AMA CAVC Reversal appeal and marks the case as ready for distribution.
- [x] Verify this case is eligible for distribution and is picked up in the priority case distribution logic (should as this is a CAVC case)
- [x] Verify the remand information is captured in the new appeals remand section

### Testing Plan
1. Open a Rails console and create a ready for CAVC appeal, note the `stream_docket_number` so you can search for the appeal in the UI or the `uuid`.
```ruby
original_appeal = FactoryBot.create(:appeal, 
                           :dispatched, 
                           associated_judge: JudgeTeam.first.judge, 
                           associated_attorney: JudgeTeam.first.attorneys.first);
FactoryBot.create_list(:request_issue, 2,                                                                                  
            :rating,                                                                                            
            :with_rating_decision_issue,                                                                           
            decision_review: original_appeal,                                                                               
            veteran_participant_id: Veteran.first.participant_id,                                                        
            contested_issue_description: "issue_description",                                                              
            notes: "issue_notes");
original_appeal.uuid
```
2. Login as a CAVC lit support user
3. Search for the dispatched appeal you just created and select it to go to Case Details: `http://localhost:3000/queue/appeals/<UUID>`
4. Click the "+ Add CAVC Remand" button in the "Post-Despatch actions" section
5. Fill out the "Add a Court Remand" form, choosing **`straight_reversal`**, and submit **without judgement & mandate dates**. (When done, repeat all these steps for **`death_dismissal`**.)
6. Copy the UUID of the newly created CAVC appeal and check out its task tree in the console
```ruby
                                          ┌─────────────────────────────────────────────────────────────────────────────┐
Appeal 1054 (E 210125-1053 Court Remand)  │ ID   │ STATUS   │ ASGN_BY │ ASGN_TO               │ UPDATED_AT              │
└── RootTask                              │ 3528 │ on_hold  │         │ Bva                   │ 2021-01-26 16:10:56 UTC │
    └── DistributionTask                  │ 3529 │ on_hold  │         │ Bva                   │ 2021-01-26 16:10:56 UTC │
        └── CavcTask                      │ 3530 │ on_hold  │         │ Bva                   │ 2021-01-26 16:10:56 UTC │
            └── MandateHoldTask           │ 3531 │ on_hold  │         │ CavcLitigationSupport │ 2021-01-26 16:10:57 UTC │
                └── TimedHoldTask         │ 3532 │ assigned │         │ CavcLitigationSupport │ 2021-01-26 16:10:56 UTC │
                                          └─────────────────────────────────────────────────────────────────────────────┘
```
or in the browser: `http://localhost:3000/task_tree/appeals/<UUID>`
7. Test the End hold early action and check the task tree
8. Play around, stress test, and find bugs early
9. Search of the original and CAVA appeals using the one docket number.
10. Check that the appeal is in the expected Queue tabs in the team view.
11. Check all the AC items.


For CASEFLOW-411, repeat all these steps except populate the form **with judgement & mandate dates**.
The resulting task tree should look like this, with no CavcTask:
```ruby
                                          ┌───────────────────────────────────────────────────────────────┐
Appeal 1052 (E 210125-1051 Court Remand)  │ ID   │ STATUS   │ ASGN_BY │ ASGN_TO │ UPDATED_AT              │
└── RootTask                              │ 3518 │ on_hold  │         │ Bva     │ 2021-01-26 16:08:32 UTC │
    └── DistributionTask                  │ 3519 │ assigned │         │ Bva     │ 2021-01-26 16:08:32 UTC │
                                          └───────────────────────────────────────────────────────────────┘
```

### User Facing Changes

 BEFORE|AFTER
 ---|---
(none) | when mandate dates are **not** populated ![image](https://user-images.githubusercontent.com/55255674/105873708-61dafb00-5fc1-11eb-9c3c-32b1c9d3919b.png)
(none) | (CASEFLOW-411) when mandate dates are populated ![image](https://user-images.githubusercontent.com/55255674/105871026-808bc280-5fbe-11eb-9d66-8ac6a6ee960f.png)


### Code Documentation Updates
